Dev Academy: Anatomy of a Smart Contract update (#375)
InterWasm/docs: dev-academy/develop-smart-contract/01-intro.md

a smart contract can expose to a user. The `execute()` function demultiplexes these different types of messages and forwards them to the appropriate message handler logic.
### Message Definition

The `ExecuteMsg` enum contains the different types of execute messages that our contract can understand.
```rust
// src/msg.rs
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum ExecuteMsg {
  Increment {},
  Reset { count: i32 },
}
```
:::note
The line `#[serde(rename_all = "snake_case")]` performs a snake_case conversion (lowercase initials with an underscore between words) on the field names before serialization and deserialization. So, we'll have `increment` and `reset` instead of `Increment` and `Reset` when serializing and deserializing across JSON formatted messages.
:::

At this point, our template contract can accept the following two types of execute messages in JSON format, embedded in a `MsgExecuteContract` message:
#### Increment

> An address can utilize the Increment function to increment the current count by 1.

Reviewer: ```suggestion
Any address can utilize the Increment function to increment the current count by 1.
```

#### Reset

The owner of the contract can reset the count to an arbitrary number. The check regarding whether a user is the contract owner is a part of the execution logic.
```json
{
  "reset": {
    "count": 5
  }
}
```
### Execution Logic

> Defined in the file `/src/contract.rs`, the `execute()` function uses Rust's pattern matching to route the received `ExecuteMsg` to the appropriate handling logic, either dispatching a `try_increment()` or a `try_reset()` function call depending on the type of message received.

Reviewer: ```suggestion
Defined in the file `/src/contract.rs`, the `execute()` function uses Rust's pattern matching to route the received `ExecuteMsg` to the appropriate handling logic, by routing to `try_increment()` or `try_reset()` function call depending on the type of message received.
```

    QueryMsg::GetCount {} => to_binary(&query_count(deps)?),
    // Return the response in byte-array format
  }
}

fn query_count(deps: Deps) -> StdResult<CountResponse> {
  let state = STATE.load(deps.storage)?;
  // Load the current contract state
  Ok(CountResponse { count: state.count })
  // Form and return a CountResponse
}
```
The query() function matches the received message with one of the QueryMsg structs defined in `/src/msg.rs` and  routes the received QueryMsg to the appropriate handling logic before returning a specific query response in byte-array format.
## Building the Contract

To build your contract, run the following command. 
```sh
cargo wasm
```
This will check for any preliminary errors and output a .wasm binary under the folder `/target/wasm32/release`.

### Optimizing your build

:::info
You will need [Docker](https://www.docker.com) installed to run this command.
:::

You will need to make sure the output WASM binary is as small as possible in order to minimize fees and stay under the size limit for the blockchain. Run the following command in the root directory of your Rust smart contract's project folder.

```sh

> sudo docker run --rm -v "$(pwd)":/code \

Reviewer: You should not need to run using sudo
